Diagnostic analyzer to identify out of range placeholders in a format string.

chborl opened the pull request:
This diagnostic analyzer checks the index values of all placeholders in a format
string. If a placeholder in the string is out of range, a warning message is displayed.

fixes bug #15856

jmarolf commented:
Thanks for the contribution @cbo425!  

@CyrusNajmabadi and @dotnet/roslyn-ide for review

jmarolf commented on src/Features/Core/Portable/Features.csproj:
```diff
@@ -731,6 +732,7 @@
   </ItemGroup>
   <ItemGroup>
     <Compile Include="CodeFixes\Suppression\TopLevelSuppressionCodeAction.cs" />
+    <None Include="ClassDiagram1.cd" />
```
Remove this please :)

Pilchie commented on src/EditorFeatures/CSharpTest/ValidateFormatString/ValidateFormatStringTests.cs:
```diff
@@ -0,0 +1,625 @@
+{
+    static void Main(string[] args)
+    {
+        string.Format(""This {1} is {2} [||]my {6} test "", ""teststring1"", ""teststring2"", ""teststring3"", ""teststring4"", ""teststring5"", ""teststring6"", ""teststring7"");
```
Shouldn't this case show the diagnostic, since the placeholder is `6`, but there is not 7th value supplied?  It isn't about the absolute number of placeholders, but the highest index right?

dpoeschl replied:
Huh? The 7th value is supplied. "teststring7"

chborl replied:
There are 7 test strings passed to be inserted into the placeholders.  The placeholders will be numbered 0 through 6.  So in this test, the {6} will not show the diagnostic.

Pilchie replied:
I fail at scrolling the PR properly.  Sorry.

Pilchie commented on src/EditorFeatures/CSharpTest/ValidateFormatString/ValidateFormatStringTests.cs:
```diff
@@ -0,0 +1,625 @@
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.ValidateFormatString)]
+        public async Task FourPlaceholders()
```
Is testing up to 4 important?  Is that because there are overloads that don't take an actual params array up to that point?

chborl replied:
There are individual overloads for 1, 2 and 3 placeholders and on the 4th one it converts it to a params object array and uses that overload.

Pilchie commented on src/EditorFeatures/CSharpTest/ValidateFormatString/ValidateFormatStringTests.cs:
```diff
@@ -0,0 +1,625 @@
+{
+    static void Main(string[] args)
+    {
+        string.Format(""{0,6}[||]"",34);
```
This (and the next few tests) don't have a space after the comma between the parameters.  Unless that's important to the test (doesn't look like it), try to format the code in quotes as if it wasn't.

@jasonmalinowski How is your copious free time projection workspace for tests coming along?

Pilchie commented on src/EditorFeatures/CSharpTest/ValidateFormatString/ValidateFormatStringTests.cs:
```diff
@@ -0,0 +1,625 @@
+}");
+        }
+        [Fact, Trait(Traits.Feature, Traits.Features.ValidateFormatString)]
+        public async Task PassingiFormatProviderandOnePlaceholderOutOfBounds()
```
Nit - I would capitalize the I here.

Pilchie commented on src/Features/CSharp/Portable/ValidateFormatString/ValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,119 @@
+namespace Microsoft.CodeAnalysis.CSharp.ValidateFormatString
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    internal sealed class ValidateFormatStringDiagnosticAnalyzer : AbstractValidateFormatStringDiagnosticAnalyzer
```
Convention is to include the language here.

Pilchie commented on src/Features/CSharp/Portable/ValidateFormatString/ValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,119 @@
+            }
+
+            SimpleNameSyntax name;
+            // When calling string.Format(...), Format will be of type MemberAccessExpressionSyntax 
```
We would tend to put the variable declaration after the comment.

Pilchie replied:
Also, technically the whole `string.Format` is the member access expression, and `Format` is the Name.

Pilchie commented on src/Features/CSharp/Portable/ValidateFormatString/ValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,119 @@
+
+                name = memberAccessExpr.Name;
+            }
+            // When using static System.String and calling Format(...), Format will be of type IdentifierNameSyntax
```
We also tend not to put a comment before the `else` - instead we'd put it inside the else block.

Pilchie commented on src/Features/CSharp/Portable/ValidateFormatString/ValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,119 @@
+            }
+
+            var hasIFormatProvider =
+                method.Parameters[0].Type.ToString().Equals(typeof(IFormatProvider).FullName);
```
What is `FullName` here?  Is it an assembly qualified name?  If so, does this work when running in a PCL/.NETCore system where IFormatProvider might be in a different assembly?

I would tend to do something like `Compilation.GetTypeByMetadataName` to find IFormatProvider.

CyrusNajmabadi replied:
avoid ".ToString" for this sort of work.  .ToString will just give you the syntax the user wrote, which may include trivia.  i.e. "System   .   IFormatProvider".

chborl replied:
@Pilchie `FullName` includes the namespace but not the assembly.  I will also switch to using `Compilation.GetTypeByMetadataName`.

Pilchie commented on src/Features/CSharp/Portable/ValidateFormatString/ValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,119 @@
+
+            var formatLiteralExpressionSyntax = (LiteralExpressionSyntax)formatArgumentSyntax.Expression;
+            formatString = formatLiteralExpressionSyntax.Token.ValueText;
+            formatStringLocation = formatArgumentSyntax.GetLocation();
```
This means that the squiggle's span will include the name in a named argument case.  I would tend to think that using just the location of the `LiteralExpressionSyntax` would be better.

CyrusNajmabadi replied:
agreed.  it's also just a suitable argument you could pass out instead of needing a string and a Location.

Pilchie commented on src/Features/CSharp/Portable/ValidateFormatString/ValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,119 @@
+        {
+            foreach (var argument in arguments)
+            {
+                if (argument.NameColon != null && !argument.NameColon.IsMissing && argument.NameColon.Name.Identifier.ValueText.Equals("format"))
```
I don't think there is a case where `NameColon` can be non-null but `IsMissing`.

However, I could imagine `Name` being missing in a case like: `string.Format( : "value")`.

CyrusNajmabadi replied:
the IsMissing check isn't necessary since you're validating htat the Identifier equals "Format".

Pilchie commented on src/Features/CSharp/Portable/ValidateFormatString/ValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,119 @@
+
+            // If using positional arguments, the format string will be the first or second
+            // argument depending on whether there Is an IFormatProvider argument.
+            return (hasIFormatProvider ? arguments[1] : arguments[0]);
```
Does something validate that there are enough aguments to index into this?  I saw tests for the no-argument case, but I don't see why we wouldn't reach here.

Can you comment on why we know this is safe?

Pilchie commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,136 @@
+                return false;
+            }
+
+            if (!symbolInfo.Symbol.ContainingSymbol.Equals(context.SemanticModel.Compilation.GetTypeByMetadataName(typeof(string).FullName)))
```
You should be able to use `WellKnownTypes` to get the `string` type more directly.

Pilchie replied:
I wouldn't try to optimize for the case where this is no string type, but in general, you can split your analyzer into a `CompilationStarted` analyzer, and lookup necessary types like this, and then avoid registering any other analyzer if the type isn't present, rather than looking it up (and failing) for every invocation expression in the project.

sharwell replied:
:bulb: For this type you can avoid the type lookup altogether.

```csharp
if (!(symbolInfo.Symbol.ContainingSymbol is INamedTypeSymbol containingType))
{
    return false;
}

if (containingType.SpecialType != SpecialType.System_String)
{
    return false;
}
```

Pilchie commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,136 @@
+            }
+
+            method = (IMethodSymbol)symbolInfo.Symbol;
+            if (method == null)
```
This can never be null (you checked `SymbolInfo.Symbol` above). `method` is either valid, and an `InvalidCastException` was thrown on the line above.

Pilchie commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,136 @@
+            // arguments and parameters, then an actual object[] must be being passed instead of
+            // passing one argument at a time. This object[] may have been created somewhere we
+            // can't analyze, so never squiggle in this case.
+            if (method.Parameters[method.Parameters.Length - 1].IsParams && numberOfArguments == method.Parameters.Length)
```
Nothing verifies that `Parameters.Length` is greater than 0 allowing this subtraction to be safe.

Pilchie commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,136 @@
+            }
+
+            // If the chosen overload ends with a params object[] but there are the same number of
+            // arguments and parameters, then an actual object[] must be being passed instead of
```
"Must" is not true here.  You could be passing a single value for the params object[].

dpoeschl replied:
If you were passing a single value for the `params object[]`, the compiler would have chosen a different overload without the `params object[]`. 

This isn't quite true for the 4.5 framework (I think), where the `IFormatProvider` didn't have overloads that explicitly took 1, 2, 3... arguments. But this works correctly on the current framework and will just ignore a failing case on an older framework, so I don't think it's a big deal.

Pilchie commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,136 @@
+
+            var formatStringWithEscapedBracketsRemoved = RemoveEscapedBrackets(formatString);
+
+            MatchCollection matches = Regex.Matches(formatStringWithEscapedBracketsRemoved, "{(.*?)}");
```
I don't *love* the use of `Regex` here (my gut tells me this is going to allocate a whole bunch of little `string`s), but I'm totally fine with this as a starting point, and we can measure and optimize after.

sharwell replied:
💭 There may be an easy way to reduce this to a single allocation (amortized) for the primary case where the format string is valid. Call `string.Format` where the format string is `formatString`, and the format args is an array containing `maxExpectedPlaceholder` (count) empty strings. If the method completes without throwing an exception, return without performing the regex analysis.

Pilchie commented on src/Features/VisualBasic/Portable/ValidateFormatString/ValidateFormatStringDiagnosticAnalyzer.vb:
```diff
@@ -0,0 +1,102 @@
+                    End If
+                End If
+            Next
+            ' If using positional arguments, the format string will be the first or second
```
Nit: add a blank line here before the comment.

Pilchie commented on src/Features/VisualBasic/Portable/ValidateFormatString/ValidateFormatStringDiagnosticAnalyzer.vb:
```diff
@@ -0,0 +1,102 @@
+            End If
+
+            formatString = simpleArguementSyntax.Expression.ToString
+            formatStringLocation = formatArgumentSyntax.GetLocation
```
As in C# - maybe want a smaller span here?

Pilchie commented on src/Features/VisualBasic/Portable/ValidateFormatString/ValidateFormatStringDiagnosticAnalyzer.vb:
```diff
@@ -0,0 +1,102 @@
+            Dim simpleArguementSyntax = TryCast(formatArgumentSyntax, SimpleArgumentSyntax)
+            If (simpleArguementSyntax Is Nothing) Then
+                Return False
+            End If
```
Given the `IsKind` check above, you should just do a `DirectCast` here.

Pilchie commented on src/Features/VisualBasic/Portable/ValidateFormatString/ValidateFormatStringDiagnosticAnalyzer.vb:
```diff
@@ -0,0 +1,102 @@
+            ValidateAndReportDiagnostic(context, hasIFormatProvider, arguments.Count, formatString, formatStringLocation)
+        End Sub
+
+        Private Function TryGetFormatStringAndLocation(arguments As SeparatedSyntaxList(Of ArgumentSyntax), hasIFormatProvider As Boolean, ByRef formatString As String, ByRef formatStringLocation As Location) As Boolean
```
Consider wrapping this onto multiple lines.

Pilchie commented on src/Features/VisualBasic/Portable/ValidateFormatString/ValidateFormatStringDiagnosticAnalyzer.vb:
```diff
@@ -0,0 +1,102 @@
+            Next
+            ' If using positional arguments, the format string will be the first or second
+            ' argument depending on whether there Is an IFormatProvider argument.
+            Return If(hasIFormatProvider, arguments(1), arguments(0))
```
Same concern as C# about accessing this collection out of bounds.

sharwell commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,136 @@
+        {
+            method = null;
+
+            if (symbolInfo.Symbol == null)
```
:bulb: Can combine the two fastest checks (non-null and expected kind) here at the start, and remove the latter from later in the method:

```csharp
if (symbolInfo.Symbol?.Kind != SymbolKind.Method)
{
    return false;
}
```

sharwell commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,136 @@
+
+        protected static bool TryGetFormatMethod(SyntaxNodeAnalysisContext context, int numberOfArguments, SymbolInfo symbolInfo, out IMethodSymbol method)
+        {
+            method = null;
```
💭 I don't see anything that checks the method name.

sharwell commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,136 @@
+            return modifiedFormatString;
+        }
+
+        protected static void ValidateAndReportDiagnostic(SyntaxNodeAnalysisContext context, bool hasIFormatProvider, int numberOfArguments, string formatString, Location formatStringLocation)
```
💭 You could use `int numberOfNonPlaceholderParameters` instead of `hasIFormatProvider` if you wanted to generalize this a bit more, though I cannot say definitively that it would be an improvement.

sharwell commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,136 @@
+            // The maximum expected placeholder value depends on the number of arguments passed.
+            // All overloads have a format string argument and possibly an IFormatProvider argument
+            // so the maximum expected placeholder value is calculated accordingly.
+            var maxExpectedPlaceholder = hasIFormatProvider ? numberOfArguments - 3 : numberOfArguments - 2;
```
:bulb: I'm not a fan of `maxExpectedPlaceholder` being an inclusive upper limit. I think it would be more clear to track `placeholderArgumentCount` instead.

sharwell commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,136 @@
+
+            if (!AllPlaceholdersInRange(formatString, maxExpectedPlaceholder, out var invalidPlaceholderText))
+            {
+                var diagnostic = Diagnostic.Create(Rule, formatStringLocation, formatString.Replace("\n", "\\n").Replace("\r", "\\r"), invalidPlaceholderText);
```
:bulb: It would be more helpful to users if the diagnostic was reported on the specific problematic placeholder(s) instead of on the entire format string.

sharwell commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,136 @@
+
+        private static string RemoveEscapedBrackets(string formatString)
+        {
+            var removeEscapedBracketsRegex = new Regex("{{|}}");
```
:memo: I'm concerned about the way this treats escaped closing brackets, specifically for the following (valid) use case.

```csharp
string.Format("{{{0}}}", "Text");
```

After the replacement, the format string being analyzed will be `__{0__}` instead of `__{0}__` (with `_` used to indicate spaces).

CyrusNajmabadi replied:
can be stored in a static member.

CyrusNajmabadi commented on src/Features/CSharp/Portable/ValidateFormatString/ValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,119 @@
+    {
+        public override void Initialize(AnalysisContext context)
+        {
+            context.RegisterSyntaxNodeAction(AnalyzeNode, SyntaxKind.InvocationExpression);
```
consider using => syntax here.

CyrusNajmabadi commented on src/Features/CSharp/Portable/ValidateFormatString/ValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,119 @@
+        private static void AnalyzeNode(SyntaxNodeAnalysisContext context)
+        {
+            var invocationExpr = (InvocationExpressionSyntax)context.Node;
+            if (invocationExpr.ArgumentList?.Arguments == null)
```
i don't believe this can happen in C#.

CyrusNajmabadi commented on src/Features/CSharp/Portable/ValidateFormatString/ValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,119 @@
+            // When calling string.Format(...), Format will be of type MemberAccessExpressionSyntax 
+            if (invocationExpr.Expression is MemberAccessExpressionSyntax memberAccessExpr)
+            {
+                if (memberAccessExpr.Name?.ToString() != "Format")
```
1. memberAccessExpr.Name cannot be null, so ?. is not necessary.  
2. Instead of calling .ToString(), call ".Identifier.ValueText".  
3. use "nameof(string.Format)".
4. Validate that the "arity" of "Name" is 0.

sharwell replied:
:memo: For those unaware, the importance of `ValueText` comes into play especially for special cases of identifier handling in C#. For example, the following are not only all valid, but they are all the same in C#:

```
string.Format("{0}", 1);
string.@Format("{0}", 1);
string.F\u006frmat("{0}", 1);
```

While the `Text` property (and `ToString`) will return the text *as written* (3 different ways here), the `ValueText` property will return the interpreted text, which in all three cases is `Format`.

CyrusNajmabadi commented on src/Features/CSharp/Portable/ValidateFormatString/ValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,119 @@
+            // When using static System.String and calling Format(...), Format will be of type IdentifierNameSyntax
+            else if (invocationExpr.Expression is IdentifierNameSyntax identifierNameSyntax)
+            {
+                if (identifierNameSyntax.ToString() != "Format")
```
same comment apply here.

CyrusNajmabadi commented on src/Features/CSharp/Portable/ValidateFormatString/ValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,119 @@
+                return;
+            }
+
+            SimpleNameSyntax name;
```
i'm not a fan of declaring a variable and initializing through multiple paths.  *Consider* doing something like:

```c#
var name = GetInvocationName();
if (name == null)
{
     return;
}
```

And extract all your logic into a helper method.

CyrusNajmabadi commented on src/Features/CSharp/Portable/ValidateFormatString/ValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,119 @@
+
+            SimpleNameSyntax name;
+            // When calling string.Format(...), Format will be of type MemberAccessExpressionSyntax 
+            if (invocationExpr.Expression is MemberAccessExpressionSyntax memberAccessExpr)
```
consider a type-switch on invocatoinExpr.Expression.

CyrusNajmabadi commented on src/Features/CSharp/Portable/ValidateFormatString/ValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,119 @@
+            return true;
+        }
+
+        private static ArgumentSyntax GetFormatStringArgument(SeparatedSyntaxList<ArgumentSyntax> arguments, bool hasIFormatProvider)
```
please wrap lines that are longer than github's width. thanks!

CyrusNajmabadi commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,136 @@
+
+        public bool OpenFileOnly(Workspace workspace) => false;
+
+        protected static bool TryGetFormatMethod(SyntaxNodeAnalysisContext context, int numberOfArguments, SymbolInfo symbolInfo, out IMethodSymbol method)
```
wrap lines longer than github please :)

CyrusNajmabadi commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,136 @@
+            var formatStringWithEscapedBracketsRemoved = RemoveEscapedBrackets(formatString);
+
+            MatchCollection matches = Regex.Matches(formatStringWithEscapedBracketsRemoved, "{(.*?)}");
+            foreach (Match match in matches)
```
Can you use var here?

dpoeschl replied:
For the `MatchCollection` yes, for the `Match` no. The latter was the subject of https://github.com/dotnet/roslyn/issues/18336

CyrusNajmabadi commented on src/Features/VisualBasic/Portable/ValidateFormatString/ValidateFormatStringDiagnosticAnalyzer.vb:
```diff
@@ -0,0 +1,102 @@
+        End Sub
+
+        Private Sub AnalyzeNode(context As SyntaxNodeAnalysisContext)
+            Dim invocationExpr = CType(context.Node, InvocationExpressionSyntax)
```
this logic looks quite similar to hte C# version.  Can we share more here (potentially using generics, or using ISyntaxFacts service)?

chborl replied:
Should I put TryGetFormatStringLiteralExpressionSyntax into ISyntaxFactsService?  Or is it too specific to go in there...

CyrusNajmabadi commented on src/Features/VisualBasic/Portable/ValidateFormatString/ValidateFormatStringDiagnosticAnalyzer.vb:
```diff
@@ -0,0 +1,102 @@
+
+        Private Sub AnalyzeNode(context As SyntaxNodeAnalysisContext)
+            Dim invocationExpr = CType(context.Node, InvocationExpressionSyntax)
+            If (invocationExpr.ArgumentList?.Arguments Is Nothing) Then
```
fyi, parens not required here for VB.

chborl commented:
tagging for review @dotnet/roslyn-ide  @Pilchie @CyrusNajmabadi @sharwell

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/ValidateFormatString/output.txt:
```diff
@@ -0,0 +1,62 @@
+ValidateFormatStringTests.cs:        string.Format(""This {0[||]} works"", ""test""); 
```
remove file.

CyrusNajmabadi commented on src/EditorFeatures/TestUtilities/Diagnostics/AbstractDiagnosticProviderBasedUserDiagnosticTest.cs:
```diff
@@ -66,39 +67,54 @@ internal virtual (DiagnosticAnalyzer, CodeFixProvider) CreateDiagnosticProviderA
+                    return diagnostics.Select(d => Tuple.Create(d, (CodeFixCollection) null));
+                  // return diagnostics.Select(d => Tuple.Create(d, new CodeFixCollection(null, TextSpan.FromBounds(0,0) , (ImmutableArray<CodeFix>), null, null, null));
+                }
                 var ids = new HashSet<string>(fixer.FixableDiagnosticIds);
```
newline after }

CyrusNajmabadi commented on src/EditorFeatures/TestUtilities/Diagnostics/AbstractDiagnosticProviderBasedUserDiagnosticTest.cs:
```diff
@@ -66,39 +67,54 @@ internal virtual (DiagnosticAnalyzer, CodeFixProvider) CreateDiagnosticProviderA
+                if (fixer == null)
+                {
+                    return diagnostics.Select(d => Tuple.Create(d, (CodeFixCollection) null));
+                  // return diagnostics.Select(d => Tuple.Create(d, new CodeFixCollection(null, TextSpan.FromBounds(0,0) , (ImmutableArray<CodeFix>), null, null, null));
```
remove

CyrusNajmabadi commented on src/EditorFeatures/TestUtilities/Diagnostics/AbstractUserDiagnosticTest.cs:
```diff
@@ -140,8 +151,16 @@ protected Document GetDocumentAndAnnotatedSpan(TestWorkspace workspace, out stri
                 {
+                    if (fixer == null)
+                    {
+                        result.Add(Tuple.Create(diagnostic, (CodeFixCollection)null));
```
why?

chborl replied:
This was added so the test infrastructure can handle diagnostics without fixers.  Otherwise, it was crashing.

CyrusNajmabadi commented on src/Features/CSharp/Portable/ValidateFormatString/CSharpValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,112 @@
+            ImmutableArray<IParameterSymbol> parameters,
+            out ITypeSymbol argsArgumentType)
+        {
+            if (!TryGetArgument(semanticModel, "args", arguments, parameters, out var argsArgument))
```
what is this magic string?

CyrusNajmabadi commented on src/Features/CSharp/Portable/ValidateFormatString/CSharpValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,112 @@
+                return false;
+            }
+
+            argsArgumentType = semanticModel.GetTypeInfo(argsArgument.Expression).Type;
```
I'm surprised most of this logic could not be shared with VB.

CyrusNajmabadi commented on src/Features/CSharp/Portable/ValidateFormatString/CSharpValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,112 @@
+            argumentSyntax = null;
+
+            // First, look for a named argument that matches
+            var namedArgument = arguments.SingleOrDefault(
```
This will throw if there are multiple.  Please add test for that as well.

CyrusNajmabadi commented on src/Features/CSharp/Portable/ValidateFormatString/CSharpValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,112 @@
+
+            // If no named argument exists, look for the named parameter
+            // and return the corresponding argument
+            var namedParameter = parameters.SingleOrDefault(p => p.Name.Equals(searchArgumentName) == true);
```
== true   is unnecessary.
SingleOrDefault may throw if tehre are multiple parameters with teh same name.

CyrusNajmabadi commented on src/Features/CSharp/Portable/ValidateFormatString/CSharpValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,112 @@
+        {
+            formatStringLiteralExpressionSyntax = null;
+
+            if (!TryGetArgument(semanticModel, "format", arguments, parameters, out var formatArgumentSyntax))
```
magic string.

CyrusNajmabadi commented on src/Features/CSharp/Portable/ValidateFormatString/CSharpValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,112 @@
+            return false;
+        }
+
+        protected override bool TryGetFormatStringLiteralExpressionSyntax(
```
Consider just reutrning the expression.  Since it's a 'Try' method callers will know it may be nll.

CyrusNajmabadi commented on src/Features/CSharp/Portable/ValidateFormatString/CSharpValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,112 @@
+        protected override SyntaxKind GetInvocationExpressionSyntaxKind()
+            => SyntaxKind.InvocationExpression;
+
+        protected override string GetLiteralExpressionSyntaxAsString(SyntaxNode syntaxNode) 
```
This does not special language knowledge.  You can just call .ToString on the node at the base layer, right?

CyrusNajmabadi commented on src/Features/CSharp/Portable/ValidateFormatString/CSharpValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,112 @@
+        protected override string GetLiteralExpressionSyntaxAsString(SyntaxNode syntaxNode) 
+            => ((LiteralExpressionSyntax)syntaxNode).Token.Text;
+
+        protected override int GetLiteralExpressionSyntaxSpanStart(SyntaxNode syntaxNode)
```
This does not need special language knowledge. You can just call syntaxNode.Span.Start at the base level.

CyrusNajmabadi commented on src/Features/Core/Portable/FeaturesResources.resx:
```diff
@@ -1247,4 +1247,10 @@ This version used in: {2}</value>
     <value>'default' expression can be simplified</value>
   </data>
+  <data name="Format_string_0_contains_invalid_placeholder_1" xml:space="preserve">
+    <value>Format string '{0}' contains invalid placeholder '{1}'</value>
```
I'm curious how this shows to the user.  The format string itself is effectively unbounded in size.  So i would not want it embedded in the error string.  I would expect to just point at the placeholder that is bad and just call that part out.

CyrusNajmabadi commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,295 @@
+        where TSyntaxKind : struct
+    {
+        private static Regex s_removeEscapedBracketsRegex = new Regex("{{");
+        private static Regex s_extractPlaceholdersRegex = new Regex("{(.*?)}");
```
Can we comment why these work?    ".*?" in particular needs some explanation.

CyrusNajmabadi commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,295 @@
+namespace Microsoft.CodeAnalysis.ValidateFormatString
+{
+    internal abstract class AbstractValidateFormatStringDiagnosticAnalyzer<TSyntaxKind> :
+        DiagnosticAnalyzer
```
If you derive from AbstractCodeStyleAnalyzer you'll get a lot of the stuff around descriptors and whatnot taken care of for you.

dpoeschl replied:
@CyrusNajmabadi But it's *not* a code style analyzer. There's nothing code style related about "your string.Format call is going to crash". Sure, there might be convenience, but there's other *meaning* behind being a code style analyzer that just doesn't fit here (most code style analyzers have options, this doesn't, etc.). I think it should be left alone. Thoughts?

CyrusNajmabadi replied:
I'm ok with that.  Note: why doesn't this have an option?  I'm thinking about the user that wants us to disable all this stuff.  Is it because we think this warning is pretty much always appropriate to show the user?  If so, i'm ok with making it non-disableable.

chborl replied:
We believe the warning should always be appropriate since it will crash if it's not fixed.  @Pilchie, what do you think?

Pilchie replied:
I'm hesitant to add any new diagnostic that doesn't have a way to be disabled.  We've even been asked for ways to disable fading of unused usings, etc.

CyrusNajmabadi commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,295 @@
+                    return;
+                }
+
+                startContext.RegisterSyntaxNodeAction(c => AnalyzeNode(c, formatProviderType),
```
wrap both parameters.

CyrusNajmabadi commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,295 @@
+            }
+
+            if (!TryGetFormatStringLiteralExpressionSyntax(
+                context.SemanticModel,
```
indent again.

CyrusNajmabadi commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,295 @@
+            if (syntaxFacts.IsSimpleMemberAccessExpression(expression))
+            {
+                var nameOfMemberAccessExpression = syntaxFacts.GetNameOfMemberAccessExpression(expression);
+                return syntaxFacts.GetIdentifierOfSimpleName(nameOfMemberAccessExpression).ValueText;
```
note: the name might be generic.  just want to make sure you're doing the right hting in that case.

chborl replied:
@CyrusNajmabadi Not sure I understand, can you give me an example?

Pilchie replied:
Consider something like: `c.Method<int>("something")`.  Does `GetIdentifierOfSimpleName` work the way you expect int that case?

CyrusNajmabadi commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,295 @@
+            return argsArray.ElementType.IsReferenceType;
+        }
+
+        internal abstract bool TryGetArgsArgumentType(
```
should be with all the other abstract mtehods.

CyrusNajmabadi commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,295 @@
+                return false;
+            }
+
+            if (!(symbolInfo.Symbol.ContainingSymbol is INamedTypeSymbol containingType))
```
fortunately, can't happen.

Pilchie replied:
Even in errors or interactive?  Can `ContainingSymbol` be null?

chborl replied:
@CyrusNajmabadi @Pilchie , to simulate an error I tried strin.Format and it bailed a few lines earlier on symbolInfo.Symbol == null, so never gets to this line of code.  Also, a quick check of other analyzers, it doesn't look like others are checking Containing Symbol for null.  Should I remove this check?

Pilchie replied:
I think we determined that as long as `symbolInfo.Symbol` is an `IMethodSymbol`, then `ContainingSymbol` will never be null.

CyrusNajmabadi commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,295 @@
+                    var diagnostic = Diagnostic.Create(
+                        Rule, 
+                        invalidPlaceholderLocation, 
+                        formatString.Replace("\n", "\\n").Replace("\r", "\\r"),
```
i don't think we'd want to embed the format string into the diagnostic.

CyrusNajmabadi commented on src/Features/VisualBasic/Portable/ValidateFormatString/VisualBasicValidateFormatStringDiagnosticAnalyzer.vb:
```diff
@@ -0,0 +1,130 @@
+            Return VisualBasicSyntaxFactsService.Instance
+        End Function
+
+        Friend Overrides Function TryGetArgsArgumentType(
```
this function looks identical to the C# version.

CyrusNajmabadi commented on src/Features/VisualBasic/Portable/ValidateFormatString/VisualBasicValidateFormatStringDiagnosticAnalyzer.vb:
```diff
@@ -0,0 +1,130 @@
+            Return True
+        End Function
+
+        Protected Overrides Function TryGetFormatStringLiteralExpressionSyntax(
```
this logic looks the same too.  I'm not sure why we can't just move this up to the base type, using the appropriate syntax facts to work agnostically over c# and vb.

CyrusNajmabadi commented:
Done with pass.

dpoeschl commented on Roslyn.sln:
```diff
@@ -1,7 +1,7 @@
 Microsoft Visual Studio Solution File, Format Version 12.00
 # Visual Studio 15
-VisualStudioVersion = 15.0.26507.0
+VisualStudioVersion = 15.0.26426.1
```
?

chborl replied:
Fix pending in next push

chborl commented:
Incorporated new feedback.  Please take a look @Pilchie @CyrusNajmabadi @sharwell .

sharwell commented on src/EditorFeatures/TestUtilities/Traits.cs:
```diff
@@ -171,6 +171,7 @@ public static class Features
             public const string TodoComments = nameof(TodoComments);
             public const string TypeInferenceService = nameof(TypeInferenceService);
             public const string Venus = nameof(Venus);
+            public const string ValidateFormatString = nameof(ValidateFormatString);
```
:bulb: In alphabetical order this would come before `Venus`.

sharwell commented on src/Features/CSharp/Portable/ValidateFormatString/CSharpValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,45 @@
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    internal class CSharpValidateFormatStringDiagnosticAnalyzer :
+        AbstractValidateFormatStringDiagnosticAnalyzer<SyntaxKind>
```
:bulb: `AbstractValidateFormatStringDiagnosticAnalyzer<SyntaxKind, ArgumentSyntax>`

Pilchie commented on src/EditorFeatures/CSharpTest/ValidateFormatString/ValidateFormatStringTests.cs:
```diff
@@ -1,22 +1,40 @@
 
 namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.ValidateFormatString
 {
+
```
Nit: Remove blank line.

Pilchie commented on src/EditorFeatures/CSharpTest/ValidateFormatString/ValidateFormatStringTests.cs:
```diff
@@ -460,10 +478,83 @@ static void Main(string[] args)
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.ValidateFormatString)]
+        public async Task WarningTuredOn()
```
Nit: Spelling of "Turned"

Pilchie commented on src/Features/CSharp/Portable/ValidateFormatString/CSharpValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -4,109 +4,42 @@
-
-            if (!TryGetArgument(semanticModel, "format", arguments, parameters, out var formatArgumentSyntax))
+            var matchingNamedArguments = arguments.Cast<ArgumentSyntax>()
+                .Where(p => p.NameColon?.Name.Identifier.ValueText.Equals(searchArgumentName) == true);
```
PERF: There are a lot of hidden allocations here: `Cast`, `Where`, `Lambada`, `Count()`, `Single()`.  Consider expanding to a foreach loop.

Do you explicitly need to handle the more than one case? If not, just return the first one you come across?

Pilchie commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -67,17 +67,32 @@ public override void Initialize(AnalysisContext context)
         private void AnalyzeNode(SyntaxNodeAnalysisContext context, INamedTypeSymbol formatProviderType)
         {
+            var optionSet = context.Options.GetDocumentOptionSetAsync(
+                context.Node.SyntaxTree, context.CancellationToken).GetAwaiter().GetResult();
```
Did we have some way that we cached the results of this?

chborl replied:
As we discussed, there isn't a great way to do this right now.  It is believed the perf hit won't be too bad since currently other analyzers utilize options in the same way.

Pilchie commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -67,17 +67,32 @@ public override void Initialize(AnalysisContext context)
         {
+            var optionSet = context.Options.GetDocumentOptionSetAsync(
+                context.Node.SyntaxTree, context.CancellationToken).GetAwaiter().GetResult();
+            if (optionSet == null)
```
You could just turn this into a `?` in the next `if`.

Pilchie commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -125,31 +141,37 @@ private void AnalyzeNode(SyntaxNodeAnalysisContext context, INamedTypeSymbol for
             }
-
-            return null;
+            
```
Nit: remove space characters.

Pilchie commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -171,11 +193,99 @@ private string GetMethodName(ISyntaxFactsService syntaxFacts, SyntaxNode express
-            out ITypeSymbol argsArgumentType);
+            out ITypeSymbol argsArgumentType)
+        {
+            var nameOfArgsParameter = "args";
```
can be `const`

Pilchie commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -171,11 +193,99 @@ private string GetMethodName(ISyntaxFactsService syntaxFacts, SyntaxNode express
+            argumentSyntax = null;
+
+            // First, look for a named argument that matches
+            var namedArguments = GetMatchingNamedArgument(arguments, searchArgumentName);
```
Nit: This is only a single argument so rename without the `s`

Pilchie commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -171,11 +193,99 @@ private string GetMethodName(ISyntaxFactsService syntaxFacts, SyntaxNode express
+            // If no named argument exists, look for the named parameter
+            // and return the corresponding argument
+            var namedParameters = parameters.Where(p => p.Name.Equals(searchArgumentName));
+
```
Nit: Remove blank line.  Also, consider rewording "named parameter" to "parameter with matching name".

Pilchie commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -171,11 +193,99 @@ private string GetMethodName(ISyntaxFactsService syntaxFacts, SyntaxNode express
+                return false;
+            }
+
+            var namedParameter = namedParameters.Single();
```
PERF: Again, consider, not using `Where()` + `Count()` + `Single()` in a hot path in an analyzer.

Pilchie commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -171,11 +193,99 @@ private string GetMethodName(ISyntaxFactsService syntaxFacts, SyntaxNode express
+            var namedParameter = namedParameters.Single();
+
+            // For the case string.Format("Test string"), there is only one argument
+            // but the compiler created an empty parameter array to bind to an overload
```
Note - this can also happen in error cases.

Pilchie commented on src/Features/VisualBasic/Portable/ValidateFormatString/VisualBasicValidateFormatStringDiagnosticAnalyzer.vb:
```diff
@@ -18,113 +15,45 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.ValidateFormatString
-                LiteralExpressionSyntax)
-
-            Return True
+            Dim argumentSyntax = TryCast(argsArgument, ArgumentSyntax)
```
You have `TryCast`, but don't check the result. Either make into a `DirectCast`, or check for failure if you think that's possible.

(The C# equivalent code does a direct cast).

Pilchie commented on src/Features/VisualBasic/Portable/ValidateFormatString/VisualBasicValidateFormatStringDiagnosticAnalyzer.vb:
```diff
@@ -18,113 +15,45 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.ValidateFormatString
-            ' argument
-            Dim namedParameter = parameters.SingleOrDefault(Function(p) p.Name.Equals(searchArgumentName))
-            If namedParameter IsNot Nothing Then
+            Return matchingArgs.Single()
```
PERF: Same concerns as in C# about all the allocations.

Pilchie commented on src/Features/VisualBasic/Portable/ValidateFormatString/VisualBasicValidateFormatStringDiagnosticAnalyzer.vb:
```diff
@@ -18,113 +15,45 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.ValidateFormatString
-        Protected Overrides Function GetLiteralExpressionSyntaxAsString(syntaxNode As SyntaxNode) As String
-            Return TryCast(syntaxNode, LiteralExpressionSyntax).Token.Text
+        Protected Overrides Function ArgumentExpressionIsStringLiteral(syntaxNode As SyntaxNode) As Boolean
+            Return TryCast(syntaxNode, ArgumentSyntax).GetArgumentExpression().IsKind(
```
Same `TryCast`/`DirectCast`.

Pilchie commented on src/Features/VisualBasic/Portable/ValidateFormatString/VisualBasicValidateFormatStringDiagnosticAnalyzer.vb:
```diff
@@ -18,113 +15,45 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.ValidateFormatString
-        Protected Overrides Function GetLiteralExpressionSyntaxSpanStart(syntaxNode As SyntaxNode) As Integer
-            Return TryCast(syntaxNode, LiteralExpressionSyntax).Token.SpanStart
+        Protected Overrides Function GetArgumentExpression(syntaxNode As SyntaxNode) As SyntaxNode
+            Return TryCast(syntaxNode, ArgumentSyntax).GetArgumentExpression
```
Same `TryCast`/`DirectCast`.

Pilchie commented on src/VisualStudio/CSharp/Impl/CSharpVSResources.resx:
```diff
@@ -516,4 +516,7 @@
     <value>When possible</value>
   </data>
+  <data name="Warn_on_invalid_string_dot_format_calls" xml:space="preserve">
+    <value>Warn on invalid string.Format method calls</value>
```
Nit: I'd change this to something else (@kuhlenh to help with words).  Maybe "Report calls to 'string.Format' with invalid placeholders"?

Is "Warn" right - is the severity always Warning?

kuhlenh replied:
* "Report uses of 'string.Format' with invalid placeholders"
* "Report invalid placeholders in 'string.Format' expressions"

@Pilchie why does this need to be an option at all? When would someone *not* want to fix this?

Pilchie replied:
When they open their 15 year old project that they are doing maintenance on, and the error list is full of reports of things that they don't care about right now.

Pilchie commented on src/VisualStudio/CSharp/Impl/Options/AdvancedOptionPageControl.xaml.cs:
```diff
@@ -7,12 +7,14 @@
 using Microsoft.CodeAnalysis.ExtractMethod;
 using Microsoft.CodeAnalysis.ImplementType;
-using Microsoft.CodeAnalysis.Shared.Options;
+using Microsoft.CodeAnalysis.Shared.Options;    
```
Nit: Added whitespace.

Pilchie commented on src/VisualStudio/CSharp/Impl/Options/AdvancedOptionPageControl.xaml.cs:
```diff
@@ -7,12 +7,14 @@
 using Microsoft.VisualStudio.LanguageServices.Implementation;
 using Microsoft.VisualStudio.LanguageServices.Implementation.Options;
 
+
```
Nit: Added blank line.

Pilchie commented on src/VisualStudio/VisualBasic/Impl/BasicVSResources.resx:
```diff
@@ -270,4 +270,7 @@
     <value>'nothing' checking:</value>
   </data>
+  <data name="Warn_on_invalid_string_dot_format_calls" xml:space="preserve">
+    <value>Warn on invalid string.Format method calls</value>
```
Same wording question as C#.  Note - capitalize "String" in VB

chborl commented:
I've incorporated the recent round of code review suggestions.  Please take a look @Pilchie @CyrusNajmabadi @sharwell @dpoeschl .

CyrusNajmabadi commented on src/EditorFeatures/TestUtilities/CodeActions/AbstractCodeActionTest.cs:
```diff
@@ -36,6 +36,11 @@ public abstract class AbstractCodeActionTest : AbstractCodeActionOrUserDiagnosti
 
+        protected override Task<ImmutableArray<Diagnostic>> GetDiagnosticsWorkerAsync(TestWorkspace workspace, TestParameters parameters)
+        {
+            return Task.FromResult(ImmutableArray<Diagnostic>.Empty);
```
I believe we have SpecializedTasks.EmptyImmutableArray.  But that's very minor.

CyrusNajmabadi commented on src/EditorFeatures/VisualBasicTest/ValidateFormatString/ValidateFormatStringTests.vb:
```diff
@@ -0,0 +1,317 @@
+                {New OptionKey(ValidateFormatStringOption.ReportInvalidPlaceholdersInStringDotFormatExpression, LanguageNames.VisualBasic), True}
+            }
+
+            Return optionsSet
```
fyi, there are help "Options" methods in the base class to make it easy to build up this dictionary

CyrusNajmabadi commented on src/Features/CSharp/Portable/ValidateFormatString/CSharpValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,47 @@
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.ValidateFormatString;
+using System.Linq;
```
sort.

CyrusNajmabadi commented on src/Features/CSharp/Portable/ValidateFormatString/CSharpValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,47 @@
+            {
+                if (argument.NameColon != null && argument.NameColon.Name.Identifier.ValueText.Equals(searchArgumentName))
+                {
+                return argument;
```
indent.

CyrusNajmabadi commented on src/Features/CSharp/Portable/ValidateFormatString/CSharpValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,47 @@
+            SyntaxNode argsArgument)
+                => semanticModel.GetTypeInfo(((ArgumentSyntax)argsArgument).Expression).Type;
+
+        protected override SyntaxNode GetMatchingNamedArgument(
```
Consider: TryGet...

CyrusNajmabadi commented on src/Features/CSharp/Portable/ValidateFormatString/CSharpValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,47 @@
+        }
+
+        protected override bool ArgumentExpressionIsStringLiteral(SyntaxNode syntaxNode)
+            => ((ArgumentSyntax)syntaxNode).Expression.IsKind(SyntaxKind.StringLiteralExpression);
```
Consider: Add a TArgumentSyntax generic parameter to AbstractValidateFormatStringDiagnosticAnalyzer so we don't have to have these casts here.

CyrusNajmabadi replied:
we have GetExpressionOfArgument and IsStringLiteralExpression on ISyntaxFactsService, so this method doesn't seem necessary.

CyrusNajmabadi replied:
(if you move to ISyntaxFacts, you don't need to have the generic parameter).

CyrusNajmabadi commented on src/Features/CSharp/Portable/ValidateFormatString/CSharpValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,47 @@
+            => ((ArgumentSyntax)syntaxNode).Expression.IsKind(SyntaxKind.StringLiteralExpression);
+
+        protected override SyntaxNode GetArgumentExpression(SyntaxNode syntaxNode)
+            => ((ArgumentSyntax)syntaxNode).Expression;
```
We have GetExpressionOfArgument on ISyntaxFacts.  So this method doesn't seem necessary.

CyrusNajmabadi commented on src/Features/Core/Portable/Features.csproj:
```diff
@@ -17,6 +17,10 @@
     </ProjectReference>
+    <ProjectReference Include="..\..\..\Compilers\CSharp\Portable\CSharpCodeAnalysis.csproj">
+      <Project>{B501A547-C911-4A05-AC6E-274A50DFF30E}</Project>
+      <Name>CSharpCodeAnalysis</Name>
```
Nooooop.

chborl replied:
Not sure how that happened, but it's gone in next PR.

CyrusNajmabadi commented on src/Features/Core/Portable/FeaturesResources.resx:
```diff
@@ -1247,4 +1247,10 @@ This version used in: {2}</value>
     <value>'default' expression can be simplified</value>
   </data>
+  <data name="Format_string_contains_invalid_placeholder_0" xml:space="preserve">
+    <value>Format string contains invalid placeholder '{0}'</value>
```
Nit: does hte {0} add anything meaningful to the user here?  They'll already get the right section squiggled, right?  So the message doesn't need this info.  Up to you.

chborl replied:
Good point, removed in next PR

CyrusNajmabadi commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,395 @@
+            => ImmutableArray.Create(Rule);
+
+        // this regex is used to remove escaped brackets from
+        // the format string before looking for valid {} pairs
```
consider, /// doc comment.

CyrusNajmabadi commented on src/Features/CSharp/Portable/ValidateFormatString/CSharpValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,47 @@
+        protected override ITypeSymbol GetArgumentExpressionType(
+            SemanticModel semanticModel, 
+            SyntaxNode argsArgument)
+                => semanticModel.GetTypeInfo(((ArgumentSyntax)argsArgument).Expression).Type;
```
Seems unneeded.  Why could we not do this in the base type once you've gotten the expression off of the argument?

chborl replied:
True, fixed in next PR.

CyrusNajmabadi commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,395 @@
+            context.RegisterCompilationStartAction(startContext =>
+            {
+                var formatProviderType = startContext.Compilation.GetTypeByMetadataName(
+                    "System.IFormatProvider");
```
consider: GetTypeByMetadataName(typeof(System.IFormatProvider).FullName))

CyrusNajmabadi commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,395 @@
+                context.Node.SyntaxTree, context.CancellationToken).GetAwaiter().GetResult();
+
+            if (optionSet?.GetOption(
+                ValidateFormatStringOption.ReportInvalidPlaceholdersInStringDotFormatExpression,
```
@jasonmalinowski Thoughts here?  It seems odd that this sort of option would be stored in .editor_config.  It's not really style, but just a decision about if we analyze this or not.  It feels like this is more of a ruleset thing...

jasonmalinowski replied:
Indeed, not sure why we need this vs. the analyzer just being turned on or off.

Pilchie replied:
This is what turns the analyzer off, since it's packaged in VS by default right now.

The alternative would be to do as Sam suggested, and move this whole thing to roslyn-analyzers.

CyrusNajmabadi commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,395 @@
+            var optionSet = context.Options.GetDocumentOptionSetAsync(
+                context.Node.SyntaxTree, context.CancellationToken).GetAwaiter().GetResult();
+
+            if (optionSet?.GetOption(
```
optionSet should not be able to be null.

CyrusNajmabadi commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,395 @@
+            }
+
+            var syntaxFacts = GetSyntaxFactsService();
+            var expression = syntaxFacts.GetExpressionOfInvocationExpression(context.Node);
```
if you have this, you no longer need your abstract method to do the same.

CyrusNajmabadi commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,395 @@
+                var nameOfMemberAccessExpression = syntaxFacts.GetNameOfMemberAccessExpression(expression);
+                return !syntaxFacts.IsGenericName(nameOfMemberAccessExpression)
+                    && syntaxFacts.GetIdentifierOfSimpleName(nameOfMemberAccessExpression).
+                    ValueText.Equals(nameof(string.Format));
```
1. wrapping here is weird.  just keep ValueText on the previous line.  
2. can just use == instead of .Equals
3. Technically wrong for VB which is case insensitive (but that's so minor i don't think we care, esp. as VB case corrects).

CyrusNajmabadi commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,395 @@
+            // IdentifierNameSyntax
+            if (syntaxFacts.IsIdentifierName(expression))
+            {
+                return syntaxFacts.GetIdentifierOfSimpleName(expression).ValueText.Equals(
```
same comments as above.

CyrusNajmabadi commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,395 @@
+                return false;
+            }
+
+            if (argsArgumentType.TypeKind != TypeKind.Array)
```
consider: mergin into previous if check.

CyrusNajmabadi replied:
consider simplifying entirely to "return argsArgumentType is IArrayTypeSymbol arrayType && arrayType.ElementType.IsReferenceType;"

chborl replied:
Yep, much more concise.  Added in the next PR.

CyrusNajmabadi commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,395 @@
+            ImmutableArray<IParameterSymbol> parameters)
+        {
+            var argsArgument = TryGetArgument(semanticModel, NameOfArgsParameter, arguments, parameters);
+            return (argsArgument == null? null : GetArgumentExpressionType(semanticModel, argsArgument));
```
1. outer parens not necessary.
2. space after ?

CyrusNajmabadi commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,395 @@
+            SeparatedSyntaxList<SyntaxNode> arguments,
+            ImmutableArray<IParameterSymbol> parameters)
+        {
+
```
remove newline.

CyrusNajmabadi commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,395 @@
+        {
+            foreach (var p in parameters)
+            {
+                if (p.Name.Equals(searchArgumentName))
```
can just do ==.

CyrusNajmabadi replied:
(bonus, will work if name is null)  (might be an issue for crazy parameters like __arglist).

chborl replied:
fixed in next PR

CyrusNajmabadi commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,395 @@
+            int numberOfArguments,
+            SymbolInfo symbolInfo)
+        {
+
```
remove newline.

CyrusNajmabadi commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,395 @@
+            {
+                string.Format(formatString, testArray);
+            }
+            catch
```
i am so torn on how i feel about this.  @Pilchie @jasonmalinowski, thoughts?  i really don't like that we're making it the norm that our code would throw exceptions expectedly as part of processing user code.

Pilchie replied:
An earlier iteration had this being detected by a set of regexes, but it was suggested to switch to this.  I think I'm in favor of this over trying to duplicate it.

jasonmalinowski replied:
If it wasn't that we regularly complain about exceptions being thrown by other teams in "regular" paths I wouldn't care too much. But we do and this will potentially make debugging really annoying.

This also makes me realize that we're presuming here the implementation of string.Format in the VS runtime matches the implementation of your target framework, which may not be the case. More generally, that makes me wish we had a way to flow this analyzer with the framework (i.e. put it in roslyn-analyzers repo instead?) so any behavior updates could be handled appropriately.

There *is* one very subtle assumption being made here which is this won't get tripped up with any format specifiers, i.e. that string always takes any format specifier. If that's ever *not* the case then this is actually broken.

The performance is also icky: it means in _correct_ code we're allocating all sorts of strings only to see if we could allocate such a string.

I can't say this approach is obviously better to me.

CyrusNajmabadi commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,395 @@
+                    var invalidPlaceholderLocation = Location.Create(
+                        context.Node.SyntaxTree,
+                        new Text.TextSpan(
+                            formatStringPosition + match.Index, 
```
you're making against formatStringWithEscapedBracketsRemoved, but then using indices to position the location inside formatString.  it seems like these locations may be incorrect.  For example, if i had ```"{{    {{   {{   {{   {1}", 0```, will the error message be in the right location?

chborl replied:
Added comment about removing the escapedBrackets and replacing with spaces to keep locations the same.

CyrusNajmabadi commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,395 @@
+        }
+
+        private static string RemoveEscapedBrackets(string formatString)
+            => s_removeEscapedBracketsRegex.Replace(formatString, "  ");
```
oh cute.  so you're removing the escapedBrackts and replacing with spaces to keep locations the same?  If so, please comment this code to explain that that's waht you're doing.

chborl replied:
Added a comment where this method is called

CyrusNajmabadi commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,395 @@
+            string textInsideBrackets, 
+            int numberOfPlaceholderArguments)
+        {
+            var placeholderIndexText = (textInsideBrackets.IndexOf(",") > 0)
```
nit: parens not necessary.

CyrusNajmabadi commented on src/Features/Core/Portable/ValidateFormatString/ValidateFormatStringOption.cs:
```diff
@@ -0,0 +1,21 @@
+﻿using System.Collections.Immutable;
```
copyright.

CyrusNajmabadi commented on src/Features/Core/Portable/ValidateFormatString/ValidateFormatStringOption.cs:
```diff
@@ -0,0 +1,21 @@
+    {
+        public static PerLanguageOption<bool> ReportInvalidPlaceholdersInStringDotFormatExpression =
+            new PerLanguageOption<bool>(nameof(ValidateFormatStringOption), nameof(ReportInvalidPlaceholdersInStringDotFormatExpression), defaultValue: true,
+                storageLocations: new RoamingProfileStorageLocation("TextEditor.%LANGUAGE%.Specific.WarnOnInvalidStringDotFormatCalls"));
```
Not stored in editor_config.  So i don't know why we neeed DocumentOptions.  David can help out here.

chborl replied:
@dpoeschl , can you help me with this?

CyrusNajmabadi commented on src/Features/Core/Portable/ValidateFormatString/ValidateFormatStringOption.cs:
```diff
@@ -0,0 +1,21 @@
+    }
+
+    [ExportOptionProvider, Shared]
+    internal class SplitStringLiteralOptionsProvider : IOptionProvider
```
the name of this type seems odd given what it contains.

CyrusNajmabadi commented on src/Features/VisualBasic/Portable/ValidateFormatString/VisualBasicValidateFormatStringDiagnosticAnalyzer.vb:
```diff
@@ -0,0 +1,53 @@
+                searchArgumentName As String) As SyntaxNode
+
+            For Each argument In arguments.Cast(Of ArgumentSyntax)
+                Dim simpleArgumentSyntax = DirectCast(argument, SimpleArgumentSyntax)
```
this will crash if it 's not a SimpleArgumentSyntax.  We need test here.

chborl replied:
@CyrusNajmabadi , I changed this to TryCast in the next PR, but I have a couple of questions for you...

If it's not SimpleArgumentSyntax, then it would have to be either RangeArgumentSyntax or OmittedArgumentSyntax (if I'm missing others let me know).  Is it possible to pass a RangeArgumentSyntax to String.Format?  And when passing OmittedArgumentSyntax, I called GetSymbolInfo on the expression earlier in the code, which returns null and exits the analyzer .  The specific test is: String.Format(,)    Is there another test case I should be running?

CyrusNajmabadi commented on src/VisualStudio/CSharp/Impl/CSharpVSResources.resx:
```diff
@@ -513,4 +513,7 @@
     <value>When possible</value>
   </data>
+  <data name="Report_invalid_placeholders_in_string_dot_format_expressions" xml:space="preserve">
+    <value>Report invalid placeholders in 'string.Format' expressions</value>
```
i think i would prefer: ... in 'string.Format' calls

instead of saying "expressions".

CyrusNajmabadi commented:
Looks great.  I just had nits and minor cleanup suggestions.

sharwell commented:
On further review, I now believe that this analyzer is more suited to roslyn-analyzers than it is to this repository. In particular, I believe this is a solid foundation for implementing [CA2241: ProvideCorrectArgumentsToFormattingMethods](https://github.com/dotnet/roslyn-analyzers/blob/3954448c5cc60dfccabcdc605d438f180a0e515c/docs/FxCopPort/RulesInventory.csv#L243). I believe the distinction is particularly important because we've run into other cases where the line between analyzers in dotnet/roslyn, dotnet/roslyn-analyzers, and other projects is blurred. With that said, keep in mind this work is not lost (far from it). Both the implementation and the tests can be used as the foundation for implementing CA2241.

chborl commented:
@Pilchie @CyrusNajmabadi @dpoeschl  

I have incorporated all of the feedback except for the following issues:

1) should we test the formatString by first running string.format even though we may knowingly throw an exception.
2) should this be moved to roslyn-analyzers, or merged in now and moved with all of the analyzers later.

333fred commented on src/EditorFeatures/CSharpTest/Diagnostics/SimplifyTypeNames/SimplifyTypeNamesTests.cs:
```diff
@@ -3755,7 +3755,6 @@ void M()
     }
 }",
                 options: Option(CodeStyleOptions.QualifyPropertyAccess, false, NotificationOption.Warning),
-                diagnosticCount: 1,
```
Why are we no longer asserting count here?

333fred replied:
And in several other tests.

---
In reply to: [123327253](https://github.com/dotnet/roslyn/pull/18718#discussion_r123327253) [](ancestors = 123327253)

chborl replied:
@dpoeschl  would you like to answer this? I know we talked about the diagnosticCount assert not actually checking anything but I can't remember the exact reason why... can the count ever be more than 1?

dpoeschl replied:
`TestDiagnosticSeverityAndCountAsync` is now `TestDiagnosticInfoAsync`. The earlier version of this method, `TestDiagnosticSeverityAndCountAsync`, checked some _basic_ stuff about all instances of a particular diagnostic, but didn't allow for testing the diagnostic message (which can vary), etc.. The new version checks more detailed information about a _single_ diagnostic. This is implicit in the name, `TestDiagnosticInfoAsync`, which also does a `.Single()` on the retrieved list of diagnostics, so anybody incorrectly using this method to attempt to ensure multiple diagnostics are handled correctly will fail at that point.

333fred replied:
It should assert that the size equals one, so that we're not getting a failure from Single throwing an exception, but rather from an explicit assertion that there is only one diagnostic.

dpoeschl replied:
We can make that change, and it will make such failures easier to diagnose (though it wouldn't take long to figure out the `.Single()` failure). I do suspect there are tons of places where our test helpers do similar things without double-checking first though... Oh well, we can make the world a slightly better place, so let's do it.

chborl replied:
Sounds good. I added the assert to verify count equals one.

333fred commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,403 @@
+                context.Node.SyntaxTree, context.CancellationToken).GetAwaiter().GetResult();
+
+            if (optionSet.GetOption(
+                ValidateFormatStringOption.ReportInvalidPlaceholdersInStringDotFormatCalls,
```
Nit: indent a level, as these are arguments to the above call.

jinujoseph commented:
I agree with @sharwell this probably belongs to roslyn-analyzers repo.
We already have [CA2241   ](https://github.com/dotnet/roslyn-analyzers/blob/4e298f6afd156d5e958d21be8afe2c3cc99fdea3/src/Microsoft.NetCore.Analyzers/Core/Runtime/ProvideCorrectArgumentsToFormattingMethods.cs) and can build on top of that. 

@dotnet/roslyn-analysis

333fred commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,403 @@
+            // IdentifierNameSyntax
+            if (syntaxFacts.IsIdentifierName(expression))
+            {
+                return syntaxFacts.GetIdentifierOfSimpleName(expression).ValueText
```
Won't this trigger if I have a method called `Format`? I think we need to do a semantic lookup here and see if the method symbol is a string.Format.

333fred replied:
Nevermind, I see we're checking this later on.

333fred commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,403 @@
+
+        protected static IMethodSymbol TryGetValidFormatMethodSymbol(
+            SyntaxNodeAnalysisContext context,
+            int numberOfArguments,
```
Unused?

chborl replied:
Good point.  Will be removed in next PR

333fred commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,403 @@
+            var numberOfArguments = arguments.Count;
+            var symbolInfo = context.SemanticModel.GetSymbolInfo(expression, context.CancellationToken);
+
+            var method = TryGetValidFormatMethodSymbol(context, numberOfArguments, symbolInfo);
```
We should make sure that this is a valid format method before enumerating all the arguments.

chborl replied:
Will move this further down in the code on next PR

333fred commented on src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,403 @@
+                return;
+            }
+
+            if (parameters.Length == 0)
```
Shouldn't a string with a placeholder and no corresponding parameter be considered an error by this analyzer? Or am I misunderstanding this analyzer?

chborl replied:
If parameters.Length is 0, that means no parameters were passed to string.Format, not even a format string.  So there's nothing the validate.

333fred replied:
Ah, right. Ignore.

333fred commented on src/EditorFeatures/CSharpTest/ValidateFormatString/ValidateFormatStringTests.cs:
```diff
@@ -0,0 +1,945 @@
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.ValidateFormatString)]
+        public async Task GenericIdentifier()
```
I'm confused as to what this test is providing? Is it just that other methods named Format shouldn't trigger the analyzer? If so, then the generic doesn't contribute much, this should be a method with the same signature as a string.Format method.

chborl replied:
In the method IsValidFormatMethod, I am calling syntaxFacts.GetIdentifierOfSimpleName.  So I need to make sure it is not GenericName syntax before I call it.  Therefore I added this test to make sure it was handling it correctly.

333fred replied:
Ok. In that case, I didn't see a test case for the scenario I mentioned, unless I missed it. Could you add one for that?

dpoeschl replied:
I'm not sure this will actually work, even if you make your own `System.String` class with a generic version of `Format`, because we're checking `containingType.SpecialType != SpecialType.System_String`, and I don't know off the top of my head which `System.String` that will pick. We can certainly find out though.

chborl replied:
I added a test for this.  And if you make your own 'System.String' class then 'containingType.SpecialType != SpecialType.System_String'

chborl commented:
Responded to latest feedback.

333fred commented on src/EditorFeatures/TestUtilities/Diagnostics/AbstractDiagnosticProviderBasedUserDiagnosticTest.cs:
```diff
@@ -64,38 +65,54 @@ internal virtual (DiagnosticAnalyzer, CodeFixProvider) CreateDiagnosticProviderA
             {
                 var diagnostics = (await GetDiagnosticsAsync(workspace, testOptions)).Where(d => d.Id == diagnosticId);
-                Assert.Equal(diagnosticCount, diagnostics.Count());
+                Assert.Equal(diagnostics.Count(), 1);
```
Expected count should be first.

dpoeschl commented:
retest windows_debug_unit32_prtest please

jrmoreno1 commented:
In VB.net the string.Format analyzer always gives an error if the paramaray is a literal array and there is more than 1 index, in c# it never gives an error, regardless of how many or few indexes there are.
